crates/wasm-encoder/src/exports.rs

```diff
@@ -48,24 +86,32 @@ impl ExportSection {
         self.num_added == 0
     }
 
-    /// Define an export.
+    /// Define an export in the export section.
     pub fn export(&mut self, name: &str, export: Export) -> &mut Self {
+        self.uses_component_exports |= match export {
+            Export::Function(_)
+            | Export::Table(_)
+            | Export::Memory(_)
+            | Export::Global(_)
+            | Export::Tag(_) => false,
+            Export::Instance(_) | Export::Module(_) | Export::AdapterFunction(_) => true,
+        };
```

Comment:
Personally I still feel that this sort of guard isn't necessary. It looks like `uses_component_exports` is purely used for an assert, and otherwise there wouldn't need to be any handling of components-vs-not.

I mentioned this before, but I still feel that `wasm-encoder` shouldn't be in the business of going the extra mile to ensure that only a valid module is emitted. There's naturally a large number of properties that aren't checked during encoding so there's a line to draw at what's checked and what isn't. I think it's reasonable to have some guarantees about the static structure via the API design and properties we can get trivially from that, but otherwise going above-and-beyond what the APIs are providing feels out of scope of `wasm-encoder` to me at least.